### DISCUSSION TITLE
Support the Navigation API

### DISCUSSION THREAD

**Discussion by loganvolkers** [remix-run/react-router#11046] [Proposals]
The Navigation API is a replacement for the History API.

It has graduated from being [a HTML Spec PR](https://github.com/whatwg/html/pull/8502) to [a part of the spec](https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-api) and landed in Chrome 110 (~ February 7th 2023, plus other Chromium-based browsers since).

Some excellent features include:

- The navigate event: a powerful way to conditionally convert cross-document navigations to same-document navigations
- entries: exposing a list of NavigationHistoryEntry as an abstraction over same-origin session history entries
- Focus reset and scroll behaviour
- Handling a 'traverse' navigation event is much more intuitive than a popstate. You can avoid reacting to it, derive whether the direction was forward or backwards by comparing entry index, and much more.

In [preliminary results from State of HTML 2023](https://github.com/web-platform-tests/interop/issues/550), the Navigation API was a somewhat common response to the freeform question "Which existing HTML features or browser APIs are you unable to use because of browser differences or lack of support?"

### Where can I learn more?
Being discussed in the interop project: https://github.com/web-platform-tests/interop/issues/435

### What should change in React Router

React Router should rely on the `navigate` event for core state changes. This will make React Router able to operate in an islands architecture of many front-end applications.

**Comment by martinszeltins** [comment]
Yes, 💯. It is finally possible to get native browser loading indicator showing when navigating between routes in a SPA. This has been long overdue.

**Comment by ryanflorence** [comment]
We were involved in very early conversations about the navigation API and are happy how it shaped up, all of our feedback seems to have been included. The moment it's practical to use it, we will.

**Reply by kiddliu** [reply]
Any schedule on this? Really love to see if Navigation API can help with the focus management #9555

**Reply by brophdawg11** [reply]
I don't know that we'll make a serious effort at adding support until it's implemented by Safari (that's when I'd consider it "practical to use it"), otherwise we just end up bloating the code since we have to send both implementations to all users to feature detect.

https://developer.mozilla.org/en-US/docs/Web/API/Navigation_API#browser_compatibility

**Reply by kiddliu** [reply]
Fully understood...

Did some quick check on [WebKit Bugzilla](https://bugs.webkit.org/show_bug.cgi?id=258384) and [PRs](https://github.com/WebKit/WebKit/pull/22865), they seems making progress to land this feature.

**Comment by enjikaka** [comment]
~In Chrome now!~

https://www.youtube.com/watch?v=bSua9vMdrNM&t=460s

**Reply by martinszeltins** [reply]
The video talks about Speculation Rules AI not Navigation API. Those are 2 different things, I believe?

**Reply by enjikaka** [reply]
Oh sorry. Should have read closer. Google tricked me to this thread... 
<img width="898" alt="Skärmavbild 2024-05-20 kl  18 38 30" src="https://github.com/remix-run/react-router/assets/877455/11509a59-b188-4444-ab76-285ed3bfd55f">

**Comment by o-t-w** [comment]
This API is landing in [Safari 26.2](https://developer.apple.com/documentation/safari-release-notes/safari-26_2-release-notes#:~:text=Added%20support%20for%20the%20Navigation%20API.) and [Firefox 147](https://groups.google.com/a/mozilla.org/g/dev-platform/c/2U3cwQgCWBY/m/9AIIL0okAgAJ).

**Comment by adamzr** [comment]
This is now baseline - available in the latest stable version of all major browsers. https://web.dev/blog/baseline-navigation-api
